### DISCUSSION TITLE
More information on how color styles work?

### DISCUSSION THREAD

**Discussion by egnor** [xonsh/xonsh#4368] [Q&A]
The [tutorial guide](https://xon.sh/customization.html#registering-custom-styles) is a start but leaves questions unanswered. (The [Pygments docs](https://pygments.org/docs/styledevelopment/#creating-own-styles) are reasonable, but xonsh seems to layer quite a bit on top.)

Some questions off the top:
- Pygments has generalization (you can define `Token.Comment` which will apply to `Token.Comment.Multiline`, `Token.Comment.Preproc`, and so on unless overridden). However it seems like the xonsh styles explicitly fill out every pygments token type, so that generalization doesn't do much, and you gotta specify all 65 of them... right?
- The xonsh style lists include both pygments token types like `Token.Literal.Number` but also a bunch of colors like `BOLD_YELLOW` and `BACKGROUND_RED`, all mixed together in one map. (The tutorial example for some reason redefines `RED` to be _green_ `"#008800"`.) Why do the color names exist? Why would one want to redefine them? Is this a sort of second level redirection? Are they related to the color names that can be used in `printx()`, `$PROMPT` and the like? Is the idea that one might tweak `RED` to be a slightly different red that fits in with the color scheme being designed?
- There's also color combos, like `WHITE__BACKGROUND_RED`. By no means are all color combos represented however! What are these for, and when would it make sense to redefine them?
- The Pygments token types are all about highlighting source code, such as Python code. And, some token types have been added for `prompt_toolkit` (`Token.PTK.*`) However, pipeline commands and arguments don't obviously fit into the taxonomy, yet they definitely get colored. Where are their colors drawn from? I made a color style that makes every single token type plain white, yet I still get colors in the commands. I _think_ this is happening in `subproc_cmd_callback` (which uses `Token.Name.Builtin` or `Token.Error`) and `subproc_arg_callback` (which uses `Token.Text` for non-files and parses `$LS_COLORS` (!) to colorize files); is that right? Is there no way to control file colors other than editing `$LS_COLORS`?

If desired, I can collect answers & take a stab at a PR to clarify some docs, or at least leave some bread crumbs. (Or maybe there's a wonderful guide somewhere that I have failed to find?)

**Comment by anki-code** [comment]
@dyuri please take a look. Thanks!

**Reply by dyuri** [reply]
First of all, I _suppose_ many of the issues come from the fact, that `xonsh` should (and do) work without `pygments`, and the coloring part is far from _clean_, so improvements are welcome :)

- Indeed, I struggled with this too, we might remove some of the tokens from the default style (`style_tools.py#180`), especially the ones set to "". I don't know why they are there, something might break...
- Yep, color names are for `printx()` and such, and they are supported in non-pygments case as well using raw ANSI escape sequences. (In case of pygments, tokens are created like `Token.Color.RED` for easier theme handling I suppose.)
- I think they are created for easy/quick pygments based colorization, composed from the basic colors defined in the previous step. You can redefine them, but I don't see any reason why you'd do that.
- This I cannot answer, but I neither like the handling of `$LS_COLORS`, especially that `xonsh` sets a default one if it's not set. (Or this was the case last time I checked.)

Clarifying the docs would be awesome, but I think we can also improve the color/token handling in general.

**Comment by egnor** [comment]
Working without pygments: That's just to reduce dependencies?

Can `printx()` and prompt strings use anything that's in the style (like "{Token.Literal.Number}blah blah{RESET}")?

I can take a stab at documenting the current state of affairs, if you can review and make sure I get it right...

**Reply by dyuri** [reply]
Yes, xonsh might work without ptk/pygments with minimal dependencies, using readline for prompt.

Currently `style_tools/partial_color_tokenize()` does not handle non-color tokens, but adding support for those seems not to be a big deal.

Having - even a draft - documentation about this part would definitely help not just using xonsh, but in the future if we want to "clean up" the styling code.
